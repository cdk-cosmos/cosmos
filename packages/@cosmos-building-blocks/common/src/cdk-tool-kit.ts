import { Construct, Stack, StackProps, CfnOutput } from '@aws-cdk/core';
import { SecureBucket } from '@cosmos-building-blocks/service';

export class CDKToolKit extends Stack {
  readonly bucket: SecureBucket;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, {
      description:
        'The CDK Tool Kit Stack. It was created by `@cosmos-building-blocks` and manages resources necessary for managing your Cloud Applications with AWS CDK.',
      ...props,
      stackName: 'CDKToolkit',
    });

    this.bucket = new SecureBucket(this, 'StagingBucket');
    new CfnOutput(this, 'BucketName', {
      value: this.bucket.bucketName,
      description: 'The name of the S3 bucket owned by the CDK toolkit stack',
    });

    new CfnOutput(this, 'BucketDomainName', {
      value: this.bucket.bucketDomainName,
      description: 'The domain name of the S3 bucket owned by the CDK toolkit stack',
    });
  }
}
