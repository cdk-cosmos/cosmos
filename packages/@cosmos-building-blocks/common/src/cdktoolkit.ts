import { Construct, Stack, StackProps, CfnOutput } from '@aws-cdk/core';
import { Bucket, BucketEncryption } from '@aws-cdk/aws-s3';

export class CDKToolkit extends Stack {
  readonly bucket: Bucket;

  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, {
      description:
        'The CDK Toolkit Stack. It was created by `@cosmos-building-blocks` and manages resources necessary for managing your Cloud Applications with AWS CDK.',
      ...props,
      stackName: 'CDKToolkit',
    });

    this.bucket = new Bucket(this, 'StagingBucket', {
      encryption: BucketEncryption.S3_MANAGED,
    });
    new CfnOutput(this, 'BucketName', {
      value: this.bucket.bucketName,
      description: 'The name of the S3 bucket owned by the CDK toolkit stack',
    });

    new CfnOutput(this, 'BucketDomainName', {
      value: this.bucket.bucketDomainName,
      description: 'The domain name of the S3 bucket owned by the CDK toolkit stack',
    });
  }
}
