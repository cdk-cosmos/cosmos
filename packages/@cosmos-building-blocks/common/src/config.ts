import { Construct, Lazy } from '@aws-cdk/core';
import { IStringParameter, StringParameter, StringParameterProps } from '@aws-cdk/aws-ssm';
import { isCrossAccount } from './utils';

export class Config extends Construct {
  readonly parent?: Config;
  readonly namespace: string;
  private states: Map<string, IParam>;

  constructor(scope: Construct, id: string, namespace: string, parent?: Config) {
    super(scope, id);

    this.parent = parent;
    this.namespace = namespace;
    this.states = new Map();
  }

  key(id: string): string {
    if (id.startsWith('/')) return id;
    const key = `${this.namespace}/${id}`;
    return this.parent ? this.parent.key(key) : '/' + key;
  }

  getState(id: string, raw = false): IParam | undefined {
    const key = this.key(id);
    if (!this.states.has(key)) {
      if (this.parent) {
        if (isCrossAccount(this, this.parent)) {
          const _state = this.parent.getState(id, true);
          if (_state) {
            const state = new Param(this, id, {
              parameterName: key,
              stringValue: _state.rawStringValue,
            });
            this.states.set(key, state);
          }
        } else {
          const state = this.parent.getState(id, raw);
          if (state) this.states.set(key, state);
        }
      }

      if (!this.states.has(key) && !raw) {
        const state = Param.fromStringParameterAttributes(this, id, {
          parameterName: key,
        }) as IParam;
        this.states.set(key, state);
      }
    }

    const state = this.states.get(key);
    if (!state) return undefined;
    if (raw && !state.rawStringValue) return undefined;
    return state;
  }

  get(id: string, raw = false): string | undefined {
    return Lazy.string({
      produce: () => {
        const state = this.getState(id, raw);
        if (!state) return undefined;
        return raw ? state.rawStringValue : state.stringValue;
      },
    });
  }

  set(id: string, value: string): void {
    const key = this.key(id);
    const state = new Param(this, id, {
      parameterName: key,
      stringValue: value,
    });
    this.states.set(key, state);
  }

  lookup(id: string): string {
    const state = this.getState(id, true);
    if (state) return StringParameter.valueFromLookup(state, state.parameterName);
    return StringParameter.valueFromLookup(this, this.key(id));
  }
}

interface IParam extends IStringParameter, Construct {
  readonly rawStringValue: string;
}

class Param extends StringParameter implements IParam {
  readonly rawStringValue: string;
  readonly parameterName: string;

  constructor(scope: Construct, id: string, props: StringParameterProps) {
    super(scope, id, props);
    this.rawStringValue = props.stringValue;
    this.parameterName = props.parameterName as string;
  }
}
