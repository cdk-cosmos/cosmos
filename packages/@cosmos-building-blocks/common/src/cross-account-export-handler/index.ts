import { CloudFormationCustomResourceEvent, CloudFormationCustomResourceResponse } from 'aws-lambda';
import { CloudFormation, ChainableTemporaryCredentials } from 'aws-sdk';
import { AssumeRoleRequest } from 'aws-sdk/clients/sts';

interface Props {
  exports: string[];
  shouldErrorIfNotFound?: boolean;
  assumeRoleArn?: string;
}

interface Attributes {
  [key: string]: string | undefined;
}

export const handler = async (
  event: CloudFormationCustomResourceEvent
): Promise<Partial<CloudFormationCustomResourceResponse> | undefined> => {
  const { exports, shouldErrorIfNotFound, assumeRoleArn } = (event.ResourceProperties as object) as Props;
  const attributes: Attributes = {};

  console.log('Event:', JSON.stringify(event, null, 2));

  if (event.RequestType === 'Delete') return;

  const credential = assumeRoleArn
    ? await getCredential({ RoleArn: assumeRoleArn, RoleSessionName: 'cross-account-stack-ref-handler' })
    : undefined;
  const cloudformation = new CloudFormation({ credentials: credential });

  const cfnExports = await getExports(cloudformation);

  for (const exp of exports) {
    const ref = cfnExports.find(x => x.Name === exp);
    if (!ref && shouldErrorIfNotFound) throw new Error(`Export ${exp} not found.`);
    attributes[exp] = ref?.Value || '';
  }

  const result = {
    Data: attributes,
  };
  console.log('Response:', JSON.stringify(result, null, 2));
  return result;
};

const getCredential = (props: AssumeRoleRequest): Promise<ChainableTemporaryCredentials> =>
  new Promise<ChainableTemporaryCredentials>((res, rej) => {
    const credential = new ChainableTemporaryCredentials({
      params: props,
    });
    credential.get(error => {
      if (error) rej(error);
      res(credential);
    });
  });

const getExports = async (cloudformation: CloudFormation, next?: string): Promise<CloudFormation.Exports> => {
  const res = await cloudformation.listExports({ NextToken: next }).promise();
  if (res.$response.error) throw res.$response.error;
  const exports: CloudFormation.Exports = [];
  if (res.Exports) exports.push(...res.Exports);
  if (res.NextToken) exports.push(...(await getExports(cloudformation, res.NextToken)));
  return exports;
};
