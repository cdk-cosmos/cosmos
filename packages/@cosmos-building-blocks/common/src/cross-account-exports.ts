import { Construct, Stack, Reference, Duration } from '@aws-cdk/core';
import { CustomResource, CustomResourceProvider } from '@aws-cdk/aws-cloudformation';
import { IFunction, Function, FunctionProps, Runtime, Code } from '@aws-cdk/aws-lambda';
import { ManagedPolicy } from '@aws-cdk/aws-iam';

export class CrossAccountExportsFn extends Function {
  constructor(scope: Construct, id: string, props?: Partial<FunctionProps>) {
    super(scope, id, {
      runtime: Runtime.NODEJS_12_X,
      code: Code.fromAsset(`${__dirname}/cross-account-stack-ref-handler.zip`),
      handler: 'index.handler',
      timeout: Duration.seconds(60),
      ...props,
    });

    // istanbul ignore next
    if (!props?.role) {
      this.role?.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('AWSCloudFormationReadOnlyAccess'));
    }
  }
}

export interface CrossAccountExportsProps {
  exports: string[];
  shouldErrorIfNotFound?: boolean;
  assumeRoleArn?: string;
  fn?: IFunction;
  alwaysUpdate?: boolean;
}

export class CrossAccountExports extends Construct {
  readonly exports: string[];
  readonly resource: CustomResource;
  readonly fn: IFunction;

  constructor(scope: Construct, id: string, props: CrossAccountExportsProps) {
    super(scope, id);

    const { exports, shouldErrorIfNotFound, assumeRoleArn, alwaysUpdate = false } = props;
    let { fn } = props;

    this.exports = exports;

    if (!fn) {
      fn = Stack.of(this).node.tryFindChild('CrossAccountExportsFn') as IFunction | undefined;
    }
    if (!fn) {
      fn = new CrossAccountExportsFn(Stack.of(this), 'CrossAccountExportsFn');
    }

    this.fn = fn;

    this.resource = new CustomResource(this, 'Resource', {
      provider: CustomResourceProvider.fromLambda(fn),
      resourceType: 'Custom::CrossAccountExports',
      properties: {
        exports: this.exports,
        shouldErrorIfNotFound,
        assumeRoleArn,
        runAt: alwaysUpdate ? new Date(Date.now()).toISOString() : undefined,
      },
    });
  }

  get(name: string): string;
  get(name?: string[]): string[];
  get(name?: string | string[]): string | string[] {
    if (typeof name === 'string') {
      return this.getRef(name).toString();
    }
    return this.getRef(name).map(x => x.toString());
  }

  getRef(name: string): Reference;
  getRef(name?: string[]): Reference[];
  getRef(name?: string | string[]): Reference | Reference[] {
    if (typeof name === 'string') {
      return this.resource.getAtt(name);
    }
    if (name) {
      return name.map(x => this.resource.getAtt(x));
    }
    return this.exports.map(x => this.resource.getAtt(x));
  }
}
