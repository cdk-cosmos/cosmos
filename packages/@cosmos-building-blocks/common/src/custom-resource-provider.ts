/* eslint-disable @typescript-eslint/no-explicit-any */
import { CfnResource } from '@aws-cdk/core';
import { CfnRole, IRole } from '@aws-cdk/aws-iam';
import { CustomResourceProvider } from '@aws-cdk/core/lib/custom-resource-provider/custom-resource-provider';

declare module '@aws-cdk/core/lib/custom-resource-provider/custom-resource-provider' {
  interface CustomResourceProviderProps {
    /**
     * Use an existing role to execute the lambda function
     */
    readonly role?: IRole;
  }
}

const getOrCreateProvider = CustomResourceProvider.getOrCreateProvider;

CustomResourceProvider.getOrCreateProvider = function(scope, uniqueid, props): CustomResourceProvider {
  const provider = getOrCreateProvider(scope, uniqueid, props);

  if (props.role) {
    const providerRole = provider.node.findChild('Role') as CfnRole;
    const providerHandler = provider.node.findChild('Handler') as CfnResource;
    provider.node.tryRemoveChild(providerRole.node.id);
    providerHandler.addPropertyOverride('Role', props.role.roleArn);
    providerHandler.addDependsOn(props.role.node.defaultChild as CfnResource);
    ((providerHandler as any).dependsOn as Set<any>).delete(providerRole);
  }

  return provider;
};
