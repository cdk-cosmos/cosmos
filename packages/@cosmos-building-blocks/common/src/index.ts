export { CDKToolkit } from './cdktoolkit';
export { CrossAccountExports, CrossAccountExportsFn, CrossAccountExportsProps } from './cross-account-exports';
