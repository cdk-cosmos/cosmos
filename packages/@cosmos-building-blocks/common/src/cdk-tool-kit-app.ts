import { App } from '@aws-cdk/core';
import { CDKToolKit } from './cdk-tool-kit';

const app = new App();
new CDKToolKit(app, 'CDKToolkit');
