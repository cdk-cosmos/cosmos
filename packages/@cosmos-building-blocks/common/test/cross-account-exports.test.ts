import '@aws-cdk/assert/jest';
import { App, Stack } from '@aws-cdk/core';
import { SynthUtils, expect as cdkExpect, countResources } from '@aws-cdk/assert';
import { Code } from '@aws-cdk/aws-lambda';
import { CrossAccountExports, CrossAccountExportsFn } from '../src';
import { Role, AnyPrincipal } from '@aws-cdk/aws-iam';

const app = new App();
const stack = new Stack(app, 'TestStack');
const exp = new CrossAccountExports(stack, 'TestExports', {
  exports: ['Test-Export'],
  fn: new CrossAccountExportsFn(stack, 'CustomFn', {
    code: Code.fromAsset(`${__dirname}/../lib/cross-account-stack-ref-handler.zip`),
    role: new Role(stack, 'TestRoleARN', { assumedBy: new AnyPrincipal() }),
  }),
});
const exp2 = new CrossAccountExports(stack, 'TestExports2', { exports: [] });
const exp3 = new CrossAccountExports(stack, 'TestExports3', { exports: [], alwaysUpdate: true });
const synth = SynthUtils.synthesize(stack);

test('should have a custom resource and a lambda fn', () => {
  expect(synth).toHaveResource('Custom::CrossAccountExports', {
    Exports: ['Test-Export'],
  });
  cdkExpect(synth).to(countResources('Custom::CrossAccountExports', 3));
  cdkExpect(synth).to(countResources('AWS::Lambda::Function', 2));
  expect(exp.get('Test-Export')).toContain('Token');
  expect(exp.get(['Test-Export'])).toHaveLength(1);
  expect(exp.get()).toHaveLength(1);
});

test('should be able to find fn in stack scope', () => {
  expect(exp.fn).not.toEqual(exp2.fn);
  expect(exp2.fn).toEqual(exp3.fn);
});

test('should match snapshot', () => {
  expect(synth.template).toMatchSnapshot();
});
