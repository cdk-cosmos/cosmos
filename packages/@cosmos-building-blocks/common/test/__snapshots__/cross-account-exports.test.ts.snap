// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Object {
  "Resources": Object {
    "CrossAccountExportsFnBB7349E9": Object {
      "DependsOn": Array [
        "CrossAccountExportsFnServiceRoleC7F2C6BD",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "MOCK",
          "S3Key": "MOCK",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CrossAccountExportsFnServiceRoleC7F2C6BD",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CrossAccountExportsFnServiceRoleC7F2C6BD": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AWSCloudFormationReadOnlyAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomFnFE03B841": Object {
      "DependsOn": Array [
        "TestRoleARN194013B0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "MOCK",
          "S3Key": "MOCK",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestRoleARN194013B0",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestExports28BDC79A7": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Exports": Array [],
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CrossAccountExportsFnBB7349E9",
            "Arn",
          ],
        },
      },
      "Type": "Custom::CrossAccountExports",
      "UpdateReplacePolicy": "Delete",
    },
    "TestExports3FD94762A": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Exports": Array [],
        "RunAt": "2020-01-01T00:00:00.000Z",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CrossAccountExportsFnBB7349E9",
            "Arn",
          ],
        },
      },
      "Type": "Custom::CrossAccountExports",
      "UpdateReplacePolicy": "Delete",
    },
    "TestExportsED6808F1": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Exports": Array [
          "Test-Export",
        ],
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomFnFE03B841",
            "Arn",
          ],
        },
      },
      "Type": "Custom::CrossAccountExports",
      "UpdateReplacePolicy": "Delete",
    },
    "TestRoleARN194013B0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
